File: Day 43/Solution.py
# 700. Search in a Binary Search Tree
# You are given the root of a binary search tree (BST) and an integer val.

# Find the node in the BST that the node's value equals val and return the subtree rooted with that node. If such a node does not exist, return null.
from typing import Optional

class TreeNode:
    def __init__(self, val: int = 0, left: 'Optional[TreeNode]' = None, right: 'Optional[TreeNode]' = None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def searchBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        while root:
            if root.val == val:
                return root
            if root.val < val:
                root = root.right
            else:
                root = root.left
        return None

# VSCode test block
if __name__ == "__main__":
    #      4
    #     / \
    #    2   7
    #   / \
    #  1   3
    root = TreeNode(4, TreeNode(2, TreeNode(1), TreeNode(3)), TreeNode(7))
    solution = Solution()
    found = solution.searchBST(root, 2)
    print(found.val if found else None)  # Expected: 2
    not_found = solution.searchBST(root, 5)
    print(not_found.val if not_found else None)  # Expected: None

File: Day 43/Solution.java
// 700. Search in a Binary Search Tree
// You are given the root of a binary search tree (BST) and an integer val.

// Find the node in the BST that the node's value equals val and return the subtree rooted with that node. If such a node does not exist, return null.
class TreeNode {
    int val;
    TreeNode left, right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class Solution {
    public TreeNode searchBST(TreeNode root, int val) {
        while (root != null) {
            if (root.val == val) return root;
            if (root.val < val) {
                root = root.right;
            } else {
                root = root.left;
            }
        }
        return null;
    }

    // Helper method for demonstration
    public static void printTreeNode(TreeNode node) {
        if (node == null) {
            System.out.println("null");
        } else {
            System.out.println(node.val);
        }
    }

    // Main for VSCode testing
    public static void main(String[] args) {
        //       4
        //      / \
        //     2   7
        //    / \
        //   1   3
        TreeNode root = new TreeNode(4, new TreeNode(2, new TreeNode(1), new TreeNode(3)), new TreeNode(7));
        Solution solution = new Solution();
        printTreeNode(solution.searchBST(root, 2)); // Expected: 2
        printTreeNode(solution.searchBST(root, 5)); // Expected: null
    }
}
